### ISSUE TITLE
A few questions on how to use

### ISSUE THREAD

**Issue by magick93** [universityofadelaide/s2i-shepherd-drupal#1]
Hi

This project looks exactly like what I am looking for. I just have a few questions:

- I have a drupal project, which is installed using composer. How do I get my project into the docker image?
- How do I add the s2i to my OpenShift environment? 
- Once this project, or the s2i is in OpenShift, am I then able to spin up my drupal application by passing in a git url, and OpenShift will then detect which s2i to use?

**Comment by pingers** [comment]
How do I get my project into the docker image?
- This happens as part of the s2i process for your project. You need a build configuration which uses the s2i builder image to build your app (which would git clone your source code and install dependencies for example).
- For the uni, we're using the Shepherd UI (a Drupal based site to manage site deployments), which generates build and deployment configurations for OpenShift. See https://github.com/universityofadelaide/shepherd. It's a work in progress, but not too far from usable for managing Drupal sites launched in OpenShift.

How do I add the s2i to my OpenShift environment?
- As far as I'm aware, s2i is included in OpenShift as of the v3 release (the change to Kubernetes).

OpenShift will then detect which s2i to use?
- Your deployment configuration specifies an image stream to use and the build config specifies which image stream to use.

Hope this helps.

**Comment by johnreytanquinco** [comment]
I am also interested to know more about this. As currently, I am trying to deploy Drupal in OpenShift. This might me an interesting solution in building the image.

**Comment by magick93** [comment]
Yes, it seems like a more 'proper' way to deploy.

s2i means source-to-image.
Its one of the features of OpenShift.

Essentially what it means is OpenShift can look at a repo, and detect what
kind of code it is.
Then, it will take that code and produce a docker image especially for
OpenShift that contains your source code.

s2i is also an image. So its a docker image that produces docker images.

**Comment by magick93** [comment]
> For the uni, we're using the Shepherd UI (a Drupal based site to manage site deployments), which generates build and deployment configurations for OpenShift. See https://github.com/universityofadelaide/shepherd. It's a work in progress, but not too far from usable for managing Drupal sites launched in OpenShift.

This sounds like a really interesting project! Unfortunately this link is broken. Can you please let me know the correct link?

> As far as I'm aware, s2i is included in OpenShift as of the v3 release (the change to Kubernetes).
Im running:

- OpenShift Master: v1.5.1+7b451fc
- Kubernetes Master: v1.5.2+43a9be4
But it doesnt appear I have the s2i utility. I will do some more research on this.

> Your deployment configuration specifies an image stream to use and the build config specifies which image stream to use.

So far the OpenShift DCs I have used have specified an Image, but not an Image Stream. 
For reference: https://docs.openshift.com/enterprise/3.0/architecture/core_concepts/builds_and_image_streams.html
One of the reasons your project is so interesting is that you use s2i - which seems to be the recommended way for doing deployments in OpenShift. But - if I understand correctly (which I may not!) image stream is different from s2i?

**Comment by magick93** [comment]
FYI - S2i can be downloaded from https://github.com/openshift/source-to-image/releases

**Comment by pingers** [comment]
Ah, we haven't made that repository public just yet... it won't be long! My bad 😢 

An image stream is essentially where your s2i built images end up... which your deployments can use. You can configure your deployments to automatically replace existing containers with the new image from the image stream.

**Comment by magick93** [comment]
Im going to write an Idiots guide to using the s2i-shepherd-drupal - so forgive me if I ask some idiotic questions. :)

**Comment by magick93** [comment]
> An image stream is essentially where your s2i built images end up... which your deployments can use. You can configure your deployments to automatically replace existing containers with the new image from the image stream.

Ok. So this relates to another question I have. Does s2i-shepherd-drupal also push to the OpenShift docker registry? Or is this handled in a separate process, eg `docker push....` ?

**Comment by pingers** [comment]
It's an all in one process... the s2i process saves your built image to the registry :)

**Comment by magick93** [comment]
Ok nice!

So where do I pass in the ulr to my OpenShift registry? Or do I need to login using `oc` before running?

Also, I would want to use another container for the database. So I would need to have env vars in the drupal container for the database connection. Are these env vars provided in the s2i?

**Comment by pingers** [comment]
Env vars would get applied as part of your deployment config - not the s2i process. s2i just builds an image... you probably shouldn't try installing a database with it... though it may be possible somehow.

https://docs.openshift.com/enterprise/3.0/architecture/core_concepts/deployments.html and 
https://docs.openshift.com/enterprise/3.0/dev_guide/deployments.html are probably helpful regarding specifying either an image or image stream for your deployments.

**Comment by bart88** [comment]
I do recommend this ebook : https://www.openshift.com/promotions/devops-with-openshift.html

**Comment by johnreytanquinco** [comment]
Hi everyone,
I have some questions in mind when trying to work on this implementation.
Please correct me if I'm mistaken with some of my inputs.

1. According to `$ s2i build test/test-app shepherd-drupal shepherd-drupal-app`,
`test/test-app` = folder where the application be placed
`shepherd-drupal` = Is this an existing Docker image you have?
`shepherd-drupal-app` = final Docker image repo tag name

2. What other parameters we can use in `$ s2i build` command? Like can we specify some `ARGS`?

**Comment by johnreytanquinco** [comment]
Hi,

I've been experimenting this project, but I have few questions in mind.
1. 
Given:
I cloned this repo, created another folder and placed my composer template/file in it then run `s2i build`. I followed the steps, from creating the builder image into running the application image.

But I used `docker run -d -p 8080:8080 shepherd-drupal-app`, since ubuntu has its default page `localhost`.

I am getting
```
This site can’t be reached
localhost refused to connect.
```
Checking: `docker ps`
```
CONTAINER ID        IMAGE                 COMMAND                CREATED             STATUS              PORTS                            NAMES
8aac67ff94b0        shepherd-drupal-app   "/usr/local/s2i/run"   20 minutes ago      Up 20 minutes       80/tcp, 0.0.0.0:8080->8080/tcp   tiny_ride

```

Is there any other logs, I can check on for monitoring?

2. Is there other parameters we can use from `s2i build test/test-app <param1> <param2>`? Like adding `ARGS`?
